ocaml/xapi-guard/lib/lwt_bounded_stream.ml

```diff
@@ -0,0 +1,48 @@
+(*
+ * Copyright (c) 2012 Citrix Systems
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *)
+
+let ( let* ) = Lwt.bind
```

Comment:
The module is needed because the function in Lwt can block the pushing thread when the queue is full. Blocking the pusher thread would block all vtpm writes, which is to be avoided at all costs

Reply:
Blocking seems the correct behaviour, what else can you do?
You could implement it like you've done below and do nothing when the limit is reached, but won't that result in data loss?

Reply:
Although it depends whether this module is used for buffering writes for xapi (we can't block and indeed would reach unbounded memory usage otherwise), or for buffering writes for disk (blocking would be correct behaviour here)

Reply:
This is used by the thread that writes the petitions to disk. In particular, it's used to notify the thread that reads from disk to push the contents to xapi. These notifications are used as an optimization to not have to keep reading the directory tree constantly to catch updates and instead read directly the files containing the updates.

Because we don't want to use an unbounded amount of memory in the queue, we limit the in-flight operations, and fall back to the worse-performing behaviour in that case.

In any case the writing thread should not be blocked because we don't want petitions from SWTPM to be blocked for too long.

Reply:
In avoidance of doubt, the elements in the bounded queue do not contain the vtpm contents, only the metadata to identify the contents on disk